<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="A manually curated arsenic functional gene database (AsgeneDB) and R package (Asgene package) are developed for rapid and accurate metagenomic analysis.">
<title>Rapid and accurate arsenic metabolic metagenomic analysis • Asgene</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Rapid and accurate arsenic metabolic metagenomic analysis">
<meta property="og:description" content="A manually curated arsenic functional gene database (AsgeneDB) and R package (Asgene package) are developed for rapid and accurate metagenomic analysis.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">Asgene</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="asgenedb-a-curated-orthology-arsenic-metabolism-gene-database-for-metagenome-annotation">AsgeneDB: A curated orthology arsenic metabolism gene database for metagenome annotation<a class="anchor" aria-label="anchor" href="#asgenedb-a-curated-orthology-arsenic-metabolism-gene-database-for-metagenome-annotation"></a>
</h1></div>
<p><br><strong>Description</strong>: A manually curated arsenic functional gene database (AsgeneDB) and R package (Asgene package) are developed for rapid and accurate metagenomic analysis.<br><br><strong>Authors</strong>: Xinwei Song, Yongguan Zhu, Yongming Luo, Jianming Xu, Bin Ma* <br></p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Arsenic (As) is a kind of toxic metal-like element widely distributed in the world. To understand the microbial community of arsenic metabolism in the environment, we developed a curated arsenic functional gene database (AsgeneDB) covering five arsenic metabolic pathways (transport, respiratory, reduction, oxidative and methylation processes), 59 arsenic biotransformation functional gene families and 414773 representative sequences. Here, protein sequences for As gene families were recruited from multiple public databases such as UniProt, NCBI RefSeq, KEGG, COG, eggNOG, arCOG and KOG. AsgeneDB covers 46 phyla and 1653 genera of bacterial, archaea and fungi. It can quickly analyze the arsenic metabolism and transformation function of microbial communities by integrating multiple lineal homology databases with high specificity, comprehensiveness, representativeness and accuracy. AsgeneDB and the associated R Package will greatly promote the study of arsenic metabolism in microbial communities in various environments.</p>
</div>
<div class="section level2">
<h2 id="asgenedb-documentation">AsgeneDB Documentation<a class="anchor" aria-label="anchor" href="#asgenedb-documentation"></a>
</h2>
<div class="section level3">
<h3 id="database-files">Database files<a class="anchor" aria-label="anchor" href="#database-files"></a>
</h3>
<div class="section level5"><h5 id="1-the-database-files-user-needs-are-already-built-into-the-r-package-asgene-therefore-when-using-asgene-r-package-to-analyze-metagenomic-data-there-is-no-need-to-download-asgenedb-separately">1. The database files user needs are already built into the R package (Asgene). Therefore, when using Asgene R package to analyze metagenomic data, there is no need to download AsgeneDB separately.<a class="anchor" aria-label="anchor" href="#1-the-database-files-user-needs-are-already-built-into-the-r-package-asgene-therefore-when-using-asgene-r-package-to-analyze-metagenomic-data-there-is-no-need-to-download-asgenedb-separately"></a>
</h5></div>
<div class="section level5"><h5 id="2-example-datasets-are-provided-as-input-and-output-to-help-users-better-understand-this-package-the-example-datasets-consist-of-three-protein-sequence-files-stored-as-testdata1rda-testdata2rda-testdata3rda">2. Example datasets are provided as input and output to help users better understand this package. The example datasets consist of three protein sequence files, stored as testdata1.rda, testdata2.rda, testdata3.rda.<a class="anchor" aria-label="anchor" href="#2-example-datasets-are-provided-as-input-and-output-to-help-users-better-understand-this-package-the-example-datasets-consist-of-three-protein-sequence-files-stored-as-testdata1rda-testdata2rda-testdata3rda"></a>
</h5></div>
<div class="section level5">
<h5 id="3-in-order-to-facilitate-users-to-use-the-asgenedb-individually-for-personalized-analysis-asgenedb-can-be-downloaded-from-httpsdatacyverseorgdav-anoniplanthomexinweiasgenedbasgenedbzip">3. In order to facilitate users to use the AsgeneDB individually for personalized analysis, AsgeneDB can be downloaded from <a href="https://data.cyverse.org/dav-anon/iplant/home/xinwei/AsgeneDB/AsgeneDB.zip" class="external-link uri">https://data.cyverse.org/dav-anon/iplant/home/xinwei/AsgeneDB/AsgeneDB.zip</a>.<a class="anchor" aria-label="anchor" href="#3-in-order-to-facilitate-users-to-use-the-asgenedb-individually-for-personalized-analysis-asgenedb-can-be-downloaded-from-httpsdatacyverseorgdav-anoniplanthomexinweiasgenedbasgenedbzip"></a>
</h5>
<div class="section level6">
<h6 id="asgenedbzip-details-includes-4-files">
<strong>AsgeneDB.zip details: includes 4 files</strong><a class="anchor" aria-label="anchor" href="#asgenedbzip-details-includes-4-files"></a>
</h6>
<ol style="list-style-type: decimal">
<li><p><strong>AsgeneDB.fa</strong>: Fasta format representative sequences obtained by clustering curated sequences at 100% sequence identity. This file can be used for “BLAST” searching arsenic genes in shotgun metagenomes.</p></li>
<li><p><strong>asgene.map</strong>: A mapping file that maps sequence IDs to gene names, only sequences belonging to arsenic gene families are included. This file is used to generate arsenic gene profiles from BLAST-like results against the database.</p></li>
<li>
<p><strong>id_gene_tax_pathway_total.csv</strong>: Species table of sequences in AsgeneDB. <br></p>
<p><strong>Columns included:</strong><br></p>
<ol style="list-style-type: decimal">
<li>Gene name (colnames:gene) <br>
</li>
<li>Corresponding arsenic metabolic pathway (colnames:pathway) <br>
</li>
<li>Taxid (colnames:taxid) <br>
</li>
<li>Protein ID (colnames:protein_id) <br>
</li>
<li>Kindom classification of species (colnames:kindom) <br>
</li>
<li>Phylum classification of species (colnames:phylum) <br>
</li>
<li>Class classification of species (colnames:class) <br><br>
</li>
<li>Order classification of species (colnames:order) <br>
</li>
<li>Family classification of species (colnames:family) <br>
</li>
<li>Genus classification of species (colnames:genus) <br>
</li>
<li>Species classification of species (colnames:species) <br>
</li>
</ol>
</li>
<li><p><strong>length.txt</strong>: The file contains the length of amino acid sequences in AsgeneDB for standardizing arsenic gene abundance statistics.</p></li>
</ol>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="dependent-tools">Dependent Tools<a class="anchor" aria-label="anchor" href="#dependent-tools"></a>
</h2>
<ol style="list-style-type: decimal"><li><strong>R Studio </strong></li></ol>
<ul>
<li>Depends: R ≥ 3.4.0</li>
<li>Imports: dplyr, seqinr</li>
</ul>
<ol start="2" style="list-style-type: decimal"><li>
<strong>database searching tools:</strong><br>
</li></ol>
<ul>
<li>usearch: <a href="https://www.drive5.com/usearch/download.html" class="external-link uri">https://www.drive5.com/usearch/download.html</a>
</li>
<li>diamond: <a href="https://github.com/bbuchfink/diamond/releases" class="external-link uri">https://github.com/bbuchfink/diamond/releases</a>
</li>
<li>blast: <a href="https://ftp.ncbi.nlm.nih.gov/blast/executables/" class="external-link uri">https://ftp.ncbi.nlm.nih.gov/blast/executables/</a>
</li>
</ul>
<!-- README.md is generated from README.Rmd. Please edit that file -->
</div>
<div class="section level2">
<h2 id="asgene-package">Asgene Package<a class="anchor" aria-label="anchor" href="#asgene-package"></a>
</h2>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>You can install the development version of Asgene from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"XinweiSong/Asgene"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<p><strong>Description</strong>:<br> we provide Asgene Package for metagenomic alignment (nucleic acid or protein sequence), subsequent gene family abundance statistics and sample abundance standardization. The database files user needs are built into the Asgene. Therefore, users only need to choose a database search tool according to their needs (e.g., USEARCH, BLAST and DIAMOND) and input three parameters (e.g., working path, search parameters of tool and filetype) to automatically analyze statistics and output statistical results. Users can select gene abundance statistics (Option: abundance) to normalize read counts per kilobase per million reads (RPKM) to eliminate differences in sequencing depth and reference sequence length between samples. In addition, if the user selects functional species statistics (Option: taxonomy), the driveing species of each arsenic metabolism gene at different classification levels in the sample can be generated automatically.</p>
</div>
<div class="section level3">
<h3 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h3>
<p>This is a basic example which shows you how to use the package:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Asgene</span><span class="op">)</span></span>
<span><span class="co">#Arsenic metabolism gene abundance analysis</span></span>
<span><span class="fu"><a href="reference/Asgene.html">Asgene</a></span><span class="op">(</span>anlysis <span class="op">=</span> <span class="st">"abundance"</span>, workdir <span class="op">=</span> <span class="st">"./"</span>, method <span class="op">=</span> <span class="st">"diamond"</span>, toolpath <span class="op">=</span> <span class="st">"./"</span>, search_parameters <span class="op">=</span> <span class="st">"-e 1e-4 -p 28 --query-cover 80 --id 50"</span>,seqtype <span class="op">=</span> <span class="st">"nucl"</span>, filetype <span class="op">=</span> <span class="st">"fasta"</span>, PE <span class="op">=</span> <span class="cn">TRUE</span> , output <span class="op">=</span> <span class="st">"./"</span><span class="op">)</span></span>
<span><span class="co">#Arsenic metabolism taxonomy analysis</span></span>
<span><span class="fu"><a href="reference/Asgene.html">Asgene</a></span><span class="op">(</span>anlysis <span class="op">=</span> <span class="st">"taxonomy"</span>, workdir <span class="op">=</span> <span class="st">"</span></span>
<span><span class="st">./"</span>, method <span class="op">=</span> <span class="st">"diamond"</span>, toolpath <span class="op">=</span> <span class="st">"./"</span>, search_parameters <span class="op">=</span> <span class="st">"-e 1e-4 -p 28 --query-cover 80 --id 50"</span>,seqtype <span class="op">=</span> <span class="st">"nucl"</span>, filetype <span class="op">=</span> <span class="st">"fasta"</span>,PE <span class="op">=</span> <span class="cn">TRUE</span>, output <span class="op">=</span> <span class="st">"./"</span><span class="op">)</span></span>
<span><span class="co">#Example datasets using</span></span>
<span><span class="fu"><a href="reference/Asgene.html">Asgene</a></span><span class="op">(</span>anlysis <span class="op">=</span> <span class="st">"abundance"</span>, workdir <span class="op">=</span> <span class="st">"./"</span>, method <span class="op">=</span> <span class="st">"diamond"</span>, toolpath <span class="op">=</span> <span class="st">"./"</span>, search_parameters <span class="op">=</span> <span class="st">"-e 1e-4 -p 28 --query-cover 80 --id 50"</span>,seqtype <span class="op">=</span> <span class="st">"prot"</span>, output <span class="op">=</span> <span class="st">"./"</span>, test.data <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/Asgene.html">Asgene</a></span><span class="op">(</span>anlysis <span class="op">=</span> <span class="st">"taxonomy"</span>, workdir <span class="op">=</span> <span class="st">"./"</span>, method <span class="op">=</span> <span class="st">"diamond"</span>, toolpath <span class="op">=</span> <span class="st">"./"</span>, search_parameters <span class="op">=</span> <span class="st">"-e 1e-4 -p 28 --query-cover 80 --id 50"</span>,seqtype <span class="op">=</span> <span class="st">"prot"</span>, output <span class="op">=</span> <span class="st">"./"</span>, test.data <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<div class="section level5">
<h5 id="output-1">Output 1<a class="anchor" aria-label="anchor" href="#output-1"></a>
</h5>
<div class="figure">
<img src="https://github.com/XinweiSong/Asgene/blob/main/results1.png" alt="Output of As metabolic gene abundance analysis" width="25%" height="30%"><p class="caption">Output of As metabolic gene abundance analysis</p>
</div>
</div>
<div class="section level5">
<h5 id="output-2">Output 2<a class="anchor" aria-label="anchor" href="#output-2"></a>
</h5>
<div class="figure">
<img src="https://github.com/XinweiSong/Asgene/blob/main/results2.png" alt="Output of As metabolic taxonomy analysis" width="25%" height="30%"><p class="caption">Output of As metabolic taxonomy analysis</p>
</div>
</div>
<div class="section level4">
<h4 id="note">
<strong>NOTE:</strong><a class="anchor" aria-label="anchor" href="#note"></a>
</h4>
<ul><li>You need to place a tab-separated file in your working path that contains the sample name and the number of sequences. Note that file extensions should not be included here. For example: <a href="https://data.cyverse.org/dav-anon/iplant/home/xinwei/AsgeneDB/sampleinfo.txt" class="external-link uri">https://data.cyverse.org/dav-anon/iplant/home/xinwei/AsgeneDB/sampleinfo.txt</a>
</li></ul>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>`use_mit_license()`, `use_gpl3_license()` or friends to pick a
    license</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Asgene</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Xinwei Song <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/YOUR-ORCID-ID" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Xinwei Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
