<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A manually curated arsenic functional gene database (AsgeneDB) and R package (Asgene package) are developed for rapid and accurate metagenomic analysis."><title>AsgeneDB: A curated orthology arsenic metabolism gene database and computational tool for metagenome annotation — Asgene • Asgene</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="AsgeneDB: A curated orthology arsenic metabolism gene database and computational tool for metagenome annotation — Asgene"><meta property="og:description" content="A manually curated arsenic functional gene database (AsgeneDB) and R package (Asgene package) are developed for rapid and accurate metagenomic analysis."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">Asgene</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>AsgeneDB: A curated orthology arsenic metabolism gene database and computational tool for metagenome annotation</h1>
      
      <div class="d-none name"><code>Asgene.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A manually curated arsenic functional gene database (AsgeneDB) and R package (Asgene package) are developed for rapid and accurate metagenomic analysis.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Asgene</span><span class="op">(</span></span>
<span>  analysis <span class="op">=</span> <span class="st">"abundance"</span>,</span>
<span>  workdir <span class="op">=</span> <span class="st">"./"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"diamond"</span>,</span>
<span>  toolpath <span class="op">=</span> <span class="st">"./"</span>,</span>
<span>  search_parameters <span class="op">=</span> <span class="st">"-e 1e-4 -p 28 --query-cover 80 --id 50"</span>,</span>
<span>  seqtype <span class="op">=</span> <span class="st">"nucl"</span>,</span>
<span>  filetype <span class="op">=</span> <span class="st">"fasta"</span>,</span>
<span>  PE <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  output <span class="op">=</span> <span class="st">"./"</span>,</span>
<span>  test.data <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>analysis</dt>
<dd><p>Choose the target you want to analyze,abundance or taxonomy; type = "character"; default = "abundance"</p></dd>


<dt>workdir</dt>
<dd><p>Specify directory for sequence file location;type = "character"; default = "./"</p></dd>


<dt>method</dt>
<dd><p>Specify the database searching tool you plan to use, currently diamond, usearch and blast are supported; type = "character"; default = "diamond"</p></dd>


<dt>toolpath</dt>
<dd><p>Specify directory for searching tool location; type = "character"; default = "./"</p></dd>


<dt>search_parameters</dt>
<dd><p>Define metagenomic comparison parameters; type = "character"; default = "-e 1e-4 -p 28 --query-cover 80 --id 50"</p></dd>


<dt>seqtype</dt>
<dd><p>Specify your sequence type, nucl or prot; type = "character"; default = "nucl"</p></dd>


<dt>filetype</dt>
<dd><p>Specify the extensions of your sequence files, e.g., fastq, fastq.gz, fasta,fasta.gz, fq, fq.gz, fa, fa.gz; type = "character"; default = "fasta"</p></dd>


<dt>PE</dt>
<dd><p>Specifies whether your metagenomic data were PE files, noting that PE files are named "_R1 with extension" or "_R2 with extension" (e.g., XINWEI_R1.fasta and XINWEI_R2.fasta); type = "logical"; default = TRUE</p></dd>


<dt>output</dt>
<dd><p>Specify the directory for the final result file; type = "character"; default = "./"</p></dd>


<dt>test.data</dt>
<dd><p>Example datasets are provided as input and output to help users better understand this package; type = "logical"; default = FALSE</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">Asgene</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">Asgene</span><span class="op">(</span>anlysis <span class="op">=</span> <span class="st">"abundance"</span>, workdir <span class="op">=</span> <span class="st">"./"</span>, method <span class="op">=</span> <span class="st">"diamond"</span>, toolpath <span class="op">=</span> <span class="st">"./"</span>, search_parameters <span class="op">=</span> <span class="st">"-e 1e-4 -p 28 --query-cover 80 --id 50"</span>, seqtype <span class="op">=</span> <span class="st">"nucl"</span>, filetype <span class="op">=</span> <span class="st">"fasta"</span>, PE <span class="op">=</span> <span class="cn">TRUE</span>, out <span class="op">=</span> <span class="st">"./"</span>, test.data <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in Asgene(anlysis = "abundance", workdir = "./", method = "diamond",     toolpath = "./", search_parameters = "-e 1e-4 -p 28 --query-cover 80 --id 50",     seqtype = "nucl", filetype = "fasta", PE = TRUE, out = "./",     test.data = FALSE):</span> unused argument (anlysis = "abundance")</span>
<span class="r-in"><span><span class="fu">Asgene</span><span class="op">(</span>anlysis <span class="op">=</span> <span class="st">"abundance"</span>, workdir <span class="op">=</span> <span class="st">"./"</span>, method <span class="op">=</span> <span class="st">"diamond"</span>, toolpath <span class="op">=</span> <span class="st">"./"</span>, search_parameters <span class="op">=</span> <span class="st">"-e 1e-4 -p 28 --query-cover 80 --id 50"</span>,seqtype <span class="op">=</span> <span class="st">"prot"</span>, output <span class="op">=</span> <span class="st">"./"</span>, test.data <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in Asgene(anlysis = "abundance", workdir = "./", method = "diamond",     toolpath = "./", search_parameters = "-e 1e-4 -p 28 --query-cover 80 --id 50",     seqtype = "prot", output = "./", test.data = TRUE):</span> unused argument (anlysis = "abundance")</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Xinwei Song.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

